docs/design.md

```diff
@@ -0,0 +1,257 @@
+# Indexer Reference Provider Design
+The goal of this reference implementation is to flesh out all the protocol details
+for data providers to be able to interact with indexer nodes. 
+
+## Overview
+- Data providers keep a local index of all the data they are providing.
+- Data providers are responsible for the advertisement of updates related to the data they are providing.
+        - Data providers must be able to uniquely identify the batches of data they are providing, and/or the
+        announcements they broadcast to the network in order to be able to serve indexer-node data ingestion requests.
+- Indexer nodes listen to advertisement from data providers. When they come across a previously unseeend
+advertisement for a data provider, they trigger a new ingestion session to sync with the provider. 
+
+## Data Structures
+Data providers and indexer nodes can use any internal representation they want for indexed data. They only
+data structures enforced by the protocol are in the interaction between the two.
+
+### Advertisements
+Advertisements are used by data providers to share updates about the data they are providing. Ideally,
+these advertisements are published in a shared pub-sub channel between providers and indexer nodes. However,
+some indexer nodes may choose to provide a dedicated endpoint to allow certain authenticated providers to push announcements and updates directly to them.
+        - `Pull` Advertisements: Indexer-nodes pull advertisement from a shared pub-sub channel.
+        - `Push` Advertisements: Authenticated providers can push advertisements to indexer-nodes.
+
+```go
+type Advertisement struct {
+        ID              AdvertisementID
+        Previous        AdvertisementID
+        Provider        peer.ID
+        Signature       []byte
+        GraphSupport    bool
+        
+}
+```
+- `ID` is the unique identifier used by the data provider to identify this update.
+- `Previous` links the current advertisement for the provider with the previous one. This allows
+providers to track the announcements they are broadcasting, and 
+indexer nodes to realize when they have missed a previous advertisement.
+- `Provider` determines the source of the advertisement.
+- `Signature` is a signature over `ID` and `Previous` to authenticate that the advertisement has been generated by `Provider`.
+- `GraphSupport` is used by providers to signal indexer nodes that they support IPLD-aware ingestion sessions (see [protocol](#protocol)).
+
+`AdvertisementID` can be chosen by data providers in the way that suit them best according to their internal
```

Comment:
can we make this a hash of the advertisement, rather than an opaque identifier so that it self-certifies the content (e.g that the `Signature` here ends up validating the contents of the advertisement)

Reply:
Great point. I added the opaque identifier as an attempt to make the protocol as general possible, but I am relax the self-certification requirement. I'll revise this.